### PR TITLE
Allow Regex in namespaceSelector.matchNames

### LINKED ISSUE #28: Allow Regex in Name Matches
Allow to use Regex expressions in the `namespaceSelector.matchNames` and `namespaceSelector.ignoreNames` (implemented in #27) fields.

This can be helpful to select multiple namespaces which don't have a common label, but a common naming scheme. For example: `namespaceSelector.ignoreNames: ^openshift-.*$`

### PR DESCRIPTION
**Pull request by ccremer**
* This will allow to include namespaces whose names match a pattern. Gives greater flexibility without having to list all included namespace one by one

Closes #28

### CODE REVIEW

File: controllers/utils.go [hunk 1]
```diff
@@ -53,16 +54,23 @@ func namespaceFromString(namespace string) v1.Namespace {
...
-func filterNamespacesByNames(names []string, namespaceList []v1.Namespace) (namespaces []v1.Namespace) {
-	nameLookup := make(map[string]bool, len(names))
+func includeNamespacesByNames(patterns []string, namespaceList []v1.Namespace) (namespaces []v1.Namespace) {
+	var compiledPatterns []*regexp.Regexp
 
-	for _, name := range names {
-		nameLookup[name] = true
+	for _, pattern := range patterns {
+		rgx, err := regexp.Compile(pattern)
+		if err == nil {
```
Comment (srueg) [inline controllers/utils.go:62]: we should probably abort here or at least log the error?

File: controllers/utils_test.go [hunk 2]
```diff
@@ -0,0 +1,39 @@
...
+		ns := "match-regex-namespace"
+
+		patterns := []string{"match-.*"}
+		sourceNs := []v1.Namespace{namespaceFromString(ns)}
+
+		result := includeNamespacesByNames(patterns, sourceNs)
+		Expect(result).To(ConsistOf(sourceNs))
+	})
+
+	It("should exclude namespace that does not match a pattern", func() {
```
Comment (srueg) [inline controllers/utils_test.go:30]: Maybe add a negative test with an illegal regex expression?

**Review by srueg — CHANGES_REQUESTED** [review]

File: controllers/utils.go (hunk 1 above)
Comment (ccremer) [reply controllers/utils.go:62]: I thought about this too. At the moment it's at least documented in the property that invalid patterns will be ignored. I see several options:
* Abort, log the error and make in the status field the `ready=false` condition (and add another condition saying that validation failed)
* Create admission webhooks (https://sdk.operatorframework.io/docs/building-operators/golang/webhooks/) that can validate new or updated SyncConfig before storing them in etcd (actively prevent updating or creating a SyncConfig that has invalid syntax). Though very interesting to get experience with, it's also probably more work and adds complexity.

File: controllers/utils_test.go (hunk 2 above)
Comment (ccremer) [reply controllers/utils_test.go:30]: I'll add one, thanks

**Comment by ccremer** [comment]
PR has grown a bit :)

I reshuffled some code around, it was becoming messy regarding status handling

File: controllers/syncconfig_status.go [hunk 3]
```diff
@@ -0,0 +1,84 @@
...
+		return err
+	}
+	r.Log.WithValues("syncCount", rc.syncCount, "deleteCount", rc.deleteCount, "failCount", rc.failCount).
+		Info("Updated SyncConfig status.", getLoggingKeysAndValuesForSyncConfig(rc.cfg)...)
+	return nil
+}
+
+// SetStatusCondition adds the given condition to the status condition of the SyncConfig. Overwrites existing conditions
+// of the same type.
+func (rc *ReconciliationContext) SetStatusCondition(condition v1alpha1.SyncConfigCondition) {
```
Comment (srueg) [inline controllers/syncconfig_status.go:32]: We can use this helper to handle this: https://github.com/kubernetes/kubernetes/pull/92717/files#diff-92b43b13f51cc268adcd402276d8cf5e8076f0a32705820a493a40b7fa0e918cR30

File: api/v1alpha1/syncconfig_types.go [hunk 4]
```diff
@@ -39,21 +39,23 @@ type (
...
+		// if at least one pattern matches. Invalid patterns will cause the sync to be cancelled and the status conditions
+		// will contain the error message.
 		MatchNames []string `json:"matchNames,omitempty"`
 	}
 
 	// SyncConfigStatus defines the observed state of SyncConfig
 	SyncConfigStatus struct {
-		// Conditions contain the states of the SyncConfig
+		// Conditions contain the states of the SyncConfig. A SyncConfig is considered Ready when at least one item has been synced.
 		Conditions []SyncConfigCondition `json:"conditions,omitempty" patchStrategy:"merge"`
```
Comment (srueg) [inline api/v1alpha1/syncconfig_types.go:53]: We can use the `metav1.Condition` instead of inventing our own.

**Review by srueg — CHANGES_REQUESTED** [review]

File: api/v1alpha1/syncconfig_types.go (hunk 4 above)
Comment (ccremer) [reply api/v1alpha1/syncconfig_types.go:53]: This is available only in K8s 1.19, we are on 1.18 and cannot upgrade due to dependency conflicts from operator sdk

**Review by srueg — APPROVED** [review]

**Pull request merged**